File: solana/programs/bridge/src/common/instructions/initialize.rs
use anchor_lang::prelude::*;

use crate::common::{
    bridge::{Bridge, Eip1559},
    Config, BRIDGE_SEED,
};

/// Accounts for the initialize instruction that sets up the bridge program's initial state.
/// This instruction creates the main bridge account for cross-chain operations between Base and
/// Solana, using the provided configuration values and initializing counters/state to zero.
#[derive(Accounts)]
pub struct Initialize<'info> {
    /// The account that pays for the transaction and bridge account creation.
    /// Must be mutable to deduct lamports for account rent.
    #[account(mut)]
    pub payer: Signer<'info>,

    /// The bridge state account being initialized.
    /// - Uses PDA with BRIDGE_SEED for deterministic address
    /// - Payer funds the account creation
    /// - Space allocated for bridge state (8-byte discriminator + Bridge::INIT_SPACE)
    #[account(
        init,
        payer = payer,
        seeds = [BRIDGE_SEED],
        bump,
        space = 8 + Bridge::INIT_SPACE
    )]
    pub bridge: Account<'info, Bridge>,

    /// The guardian account that will have administrative authority over the bridge.
    /// Must be a signer to prove ownership of the guardian key. The payer and guardian
    /// may be distinct signers.
    pub guardian: Signer<'info>,

    /// System program required for creating new accounts.
    /// Used internally by Anchor for account initialization.
    pub system_program: Program<'info, System>,
}

/// Initializes the `Bridge` state account with the provided configs, sets the guardian to the
/// provided signer, starts unpaused, zeros counters, sets the EIP-1559 base fee to
/// `eip1559_config.minimum_base_fee`, and records the current timestamp as the window start.
pub fn initialize_handler(ctx: Context<Initialize>, cfg: Config) -> Result<()> {
    let current_timestamp = Clock::get()?.unix_timestamp;
    let minimum_base_fee = cfg.eip1559_config.minimum_base_fee;

    cfg.validate()?;

    *ctx.accounts.bridge = Bridge {
        base_block_number: 0,
        nonce: 0,
        guardian: ctx.accounts.guardian.key(),
        paused: false, // Initialize bridge as unpaused
        eip1559: Eip1559 {
            config: cfg.eip1559_config,
            current_base_fee: minimum_base_fee,
            current_window_gas_used: 0,
            window_start_time: current_timestamp,
        },
        gas_config: cfg.gas_config,
        protocol_config: cfg.protocol_config,
        buffer_config: cfg.buffer_config,
        partner_oracle_config: cfg.partner_oracle_config,
        base_oracle_config: cfg.base_oracle_config,
    };

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    use anchor_lang::{
        solana_program::{
            example_mocks::solana_sdk::system_program, instruction::Instruction,
            native_token::LAMPORTS_PER_SOL,
        },
        InstructionData,
    };
    use litesvm::LiteSVM;
    use solana_keypair::Keypair;
    use solana_message::Message;
    use solana_signer::Signer;
    use solana_transaction::Transaction;

    use crate::{
        accounts,
        common::{
            bridge::{BufferConfig, Eip1559Config, GasConfig, PartnerOracleConfig, ProtocolConfig},
            BaseOracleConfig,
        },
        instruction::Initialize,
        test_utils::mock_clock,
        ID,
    };

    const TEST_TIMESTAMP: i64 = 1747440000; // May 16th, 2025

    fn setup_env() -> (LiteSVM, Keypair, Keypair, Pubkey, Vec<AccountMeta>) {
        let mut svm = LiteSVM::new();
        svm.add_program_from_file(ID, "../../target/deploy/bridge.so")
            .unwrap();

        // Create test accounts
        let payer = Keypair::new();
        let payer_pk = payer.pubkey();
        svm.airdrop(&payer_pk, LAMPORTS_PER_SOL).unwrap();

        // Create guardian keypair
        let guardian = Keypair::new();
        let guardian_pk = guardian.pubkey();

        // Mock the clock to ensure we get a proper timestamp
        mock_clock(&mut svm, TEST_TIMESTAMP);

        // Find the PDAs
        let bridge_pda = Pubkey::find_program_address(&[BRIDGE_SEED], &ID).0;

        // Build the Initialize instruction accounts
        let accounts = accounts::Initialize {
            payer: payer_pk,
            bridge: bridge_pda,
            guardian: guardian_pk,
            system_program: system_program::ID,
        }
        .to_account_metas(None);

        (svm, payer, guardian, bridge_pda, accounts)
    }

    #[test]
    fn test_initialize_handler() {
        let (mut svm, payer, guardian, bridge_pda, accounts) = setup_env();
        let payer_pk = payer.pubkey();
        let guardian_pk = guardian.pubkey();

        // Build the Initialize instruction (no guardian parameter needed)
        let gas_fee_receiver = Pubkey::new_unique();
        let ix = Instruction {
            program_id: ID,
            accounts,
            data: Initialize {
                cfg: Config {
                    eip1559_config: Eip1559Config::test_new(),
                    gas_config: GasConfig::test_new(gas_fee_receiver),
                    protocol_config: ProtocolConfig::test_new(),
                    buffer_config: BufferConfig::test_new(),
                    partner_oracle_config: PartnerOracleConfig::default(),
                    base_oracle_config: BaseOracleConfig::test_new(),
                },
            }
            .data(),
        };

        // Build the transaction with both payer and guardian as signers
        let tx = Transaction::new(
            &[&payer, &guardian],
            Message::new(&[ix], Some(&payer_pk)),
            svm.latest_blockhash(),
        );

        // Send the transaction
        svm.send_transaction(tx)
            .expect("Failed to send transaction");

        // Assert the Bridge account state is correctly initialized
        let bridge = svm.get_account(&bridge_pda).unwrap();
        assert_eq!(bridge.owner, ID);
        let bridge = Bridge::try_deserialize(&mut &bridge.data[..]).unwrap();

        // Assert the Bridge state is correctly initialized
        assert_eq!(
            bridge,
            Bridge {
                base_block_number: 0,
                nonce: 0,
                guardian: guardian_pk,
                paused: false,
                eip1559: Eip1559 {
                    config: Eip1559Config::test_new(),
                    current_base_fee: 1,
                    current_window_gas_used: 0,
                    window_start_time: TEST_TIMESTAMP,
                },
                gas_config: GasConfig::test_new(gas_fee_receiver),
                protocol_config: ProtocolConfig::test_new(),
                buffer_config: BufferConfig::test_new(),
                partner_oracle_config: PartnerOracleConfig::default(),
                base_oracle_config: BaseOracleConfig::test_new(),
            }
        );
    }

    #[test]
    fn test_initialize_partner_threshold_too_high_fails() {
        let (mut svm, payer, guardian, _bridge_pda, accounts) = setup_env();
        let payer_pk = payer.pubkey();

        // Build the Initialize instruction with an invalid partner threshold (> 5)
        let gas_fee_receiver = Pubkey::new_unique();
        let ix = Instruction {
            program_id: ID,
            accounts,
            data: Initialize {
                cfg: Config {
                    eip1559_config: Eip1559Config::test_new(),
                    gas_config: GasConfig::test_new(gas_fee_receiver),
                    protocol_config: ProtocolConfig::test_new(),
                    buffer_config: BufferConfig::test_new(),
                    partner_oracle_config: PartnerOracleConfig {
                        required_threshold: 6,
                    },
                    base_oracle_config: BaseOracleConfig::test_new(),
                },
            }
            .data(),
        };

        // Build the transaction with both payer and guardian as signers
        let tx = Transaction::new(
            &[&payer, &guardian],
            Message::new(&[ix], Some(&payer_pk)),
            svm.latest_blockhash(),
        );

        // Send the transaction and expect failure
        let result = svm.send_transaction(tx);
        assert!(result.is_err());
    }

    #[test]
    fn test_initialize_base_oracle_threshold_zero_fails() {
        let (mut svm, payer, guardian, _bridge_pda, accounts) = setup_env();
        let payer_pk = payer.pubkey();

        // Build the Initialize instruction with an invalid base oracle threshold (== 0)
        let gas_fee_receiver = Pubkey::new_unique();
        let mut base_oracle_config = BaseOracleConfig::test_new();
        base_oracle_config.threshold = 0;

        let ix = Instruction {
            program_id: ID,
            accounts,
            data: Initialize {
                cfg: Config {
                    eip1559_config: Eip1559Config::test_new(),
                    gas_config: GasConfig::test_new(gas_fee_receiver),
                    protocol_config: ProtocolConfig::test_new(),
                    buffer_config: BufferConfig::test_new(),
                    partner_oracle_config: PartnerOracleConfig::default(),
                    base_oracle_config,
                },
            }
            .data(),
        };

        // Build the transaction with both payer and guardian as signers
        let tx = Transaction::new(
            &[&payer, &guardian],
            Message::new(&[ix], Some(&payer_pk)),
            svm.latest_blockhash(),
        );

        // Send the transaction and expect failure
        let result = svm.send_transaction(tx);
        assert!(result.is_err());
    }

    #[test]
    fn test_initialize_base_oracle_threshold_gt_signer_count_fails() {
        let (mut svm, payer, guardian, _bridge_pda, accounts) = setup_env();
        let payer_pk = payer.pubkey();

        // Build the Initialize instruction with threshold > signer_count
        let gas_fee_receiver = Pubkey::new_unique();
        let mut base_oracle_config = BaseOracleConfig::test_new();
        base_oracle_config.threshold = base_oracle_config.signer_count + 1; // 2 > 1

        let ix = Instruction {
            program_id: ID,
            accounts,
            data: Initialize {
                cfg: Config {
                    eip1559_config: Eip1559Config::test_new(),
                    gas_config: GasConfig::test_new(gas_fee_receiver),
                    protocol_config: ProtocolConfig::test_new(),
                    buffer_config: BufferConfig::test_new(),
                    partner_oracle_config: PartnerOracleConfig::default(),
                    base_oracle_config,
                },
            }
            .data(),
        };

        // Build the transaction with both payer and guardian as signers
        let tx = Transaction::new(
            &[&payer, &guardian],
            Message::new(&[ix], Some(&payer_pk)),
            svm.latest_blockhash(),
        );

        // Send the transaction and expect failure
        let result = svm.send_transaction(tx);
        assert!(result.is_err());
    }

    #[test]
    fn test_initialize_base_oracle_signer_count_exceeds_array_len_fails() {
        let (mut svm, payer, guardian, _bridge_pda, accounts) = setup_env();
        let payer_pk = payer.pubkey();

        // Build the Initialize instruction with signer_count > signers.len()
        let gas_fee_receiver = Pubkey::new_unique();
        let mut base_oracle_config = BaseOracleConfig::test_new();
        base_oracle_config.signer_count = (base_oracle_config.signers.len() + 1) as u8; // exceed fixed array length
        base_oracle_config.threshold = 1; // keep valid threshold

        let ix = Instruction {
            program_id: ID,
            accounts,
            data: Initialize {
                cfg: Config {
                    eip1559_config: Eip1559Config::test_new(),
                    gas_config: GasConfig::test_new(gas_fee_receiver),
                    protocol_config: ProtocolConfig::test_new(),
                    buffer_config: BufferConfig::test_new(),
                    partner_oracle_config: PartnerOracleConfig::default(),
                    base_oracle_config,
                },
            }
            .data(),
        };

        // Build the transaction with both payer and guardian as signers
        let tx = Transaction::new(
            &[&payer, &guardian],
            Message::new(&[ix], Some(&payer_pk)),
            svm.latest_blockhash(),
        );

        // Send the transaction and expect failure
        let result = svm.send_transaction(tx);
        assert!(result.is_err());
    }

    #[test]
    fn test_initialize_base_oracle_duplicate_signers_fails() {
        let (mut svm, payer, guardian, _bridge_pda, accounts) = setup_env();
        let payer_pk = payer.pubkey();

        // Build the Initialize instruction with duplicate signer addresses among the provided entries
        let gas_fee_receiver = Pubkey::new_unique();
        let mut base_oracle_config = BaseOracleConfig::test_new();
        base_oracle_config.signer_count = 2; // consider first two entries
        base_oracle_config.threshold = 1; // keep valid threshold

        // Force a duplicate among the first `signer_count` addresses
        base_oracle_config.signers[1] = base_oracle_config.signers[0];

        let ix = Instruction {
            program_id: ID,
            accounts,
            data: Initialize {
                cfg: Config {
                    eip1559_config: Eip1559Config::test_new(),
                    gas_config: GasConfig::test_new(gas_fee_receiver),
                    protocol_config: ProtocolConfig::test_new(),
                    buffer_config: BufferConfig::test_new(),
                    partner_oracle_config: PartnerOracleConfig::default(),
                    base_oracle_config,
                },
            }
            .data(),
        };

        // Build the transaction with both payer and guardian as signers
        let tx = Transaction::new(
            &[&payer, &guardian],
            Message::new(&[ix], Some(&payer_pk)),
            svm.latest_blockhash(),
        );

        // Send the transaction and expect failure due to duplicate signer detection
        let result = svm.send_transaction(tx);
        assert!(result.is_err());
    }
}

File: solana/scripts/onchain/initialize.ts
import {
  createSignerFromKeyPair,
  generateKeyPair,
  getProgramDerivedAddress,
} from "@solana/kit";
import { SYSTEM_PROGRAM_ADDRESS } from "@solana-program/system";

import {
  fetchBridge,
  getInitializeInstruction,
} from "../../clients/ts/generated/bridge";
import { getIdlConstant } from "../utils/idl-constants";
import { CONSTANTS } from "../constants";
import { buildAndSendTransaction, getPayer, getRpc } from "./utils/transaction";
import { getTarget } from "../utils/argv";
import { toBytes, toHex } from "viem";

async function main() {
  const target = getTarget();
  const constants = CONSTANTS[target];
  const rpc = getRpc(target);
  const payer = await getPayer();

  console.log("=".repeat(40));
  console.log(`Target: ${target}`);
  console.log(`RPC URL: ${constants.rpcUrl}`);
  console.log(`Bridge: ${constants.solanaBridge}`);
  console.log(`Payer: ${payer.address}`);
  console.log("=".repeat(40));
  console.log("");

  // Derive the bridge PDA.
  const [bridgeAddress] = await getProgramDerivedAddress({
    programAddress: constants.solanaBridge,
    seeds: [Buffer.from(getIdlConstant("BRIDGE_SEED"))],
  });

  // TODO: Use the real guardian.
  const guardian = await createSignerFromKeyPair(await generateKeyPair());
  const eip1559Config = {
    target: 5_000_000,
    denominator: 2,
    windowDurationSeconds: 1,
    minimumBaseFee: 1,
  };
  const gasConfig = {
    gasPerCall: 125_000,
    gasCostScaler: 1_000_000,
    gasCostScalerDp: 1_000_000,
    gasFeeReceiver: payer.address,
  };
  const protocolConfig = {
    blockIntervalRequirement: 300,
  };
  const bufferConfig = {
    maxCallBufferSize: 8 * 1024,
  };
  const baseOracleConfig = {
    threshold: 2,
    signerCount: 2,
    signers: [
      toBytes(constants.solanaEvmLocalKey),
      toBytes(constants.solanaEvmKeychainKey),
      ...Array.from(
        { length: 14 },
        () => new Uint8Array(toBytes(constants.solanaEvmLocalKey).length)
      ),
    ],
  };
  const partnerOracleConfig = {
    requiredThreshold: 3,
  };

  // Build the instruction.
  console.log("🛠️  Building instruction...");
  const ix = getInitializeInstruction(
    {
      // Accounts
      payer: payer,
      bridge: bridgeAddress,
      systemProgram: SYSTEM_PROGRAM_ADDRESS,

      // Arguments
      guardian,
      eip1559Config,
      gasConfig,
      protocolConfig,
      bufferConfig,
      baseOracleConfig,
      partnerOracleConfig,
    },
    { programAddress: constants.solanaBridge }
  );

  // Send the transaction.
  console.log("🚀 Sending transaction...");
  await buildAndSendTransaction(target, [ix]);
  console.log("✅ Done!");

  console.log("Confirming bridge configuration...");
  const bridgeData = await fetchBridge(rpc, bridgeAddress);

  // EIP1559 confirmation
  if (bridgeData.data.guardian !== guardian.address) {
    throw new Error("Guardian mismatch!");
  }
  if (bridgeData.data.eip1559.config.target !== BigInt(eip1559Config.target)) {
    throw new Error("EIP-1559 target mismatch!");
  }
  if (
    bridgeData.data.eip1559.config.denominator !==
    BigInt(eip1559Config.denominator)
  ) {
    throw new Error("EIP-1559 denominator mismatch!");
  }
  if (
    bridgeData.data.eip1559.config.windowDurationSeconds !==
    BigInt(eip1559Config.windowDurationSeconds)
  ) {
    throw new Error("EIP-1559 windowDurationSeconds mismatch!");
  }
  if (
    bridgeData.data.eip1559.config.minimumBaseFee !==
    BigInt(eip1559Config.minimumBaseFee)
  ) {
    throw new Error("EIP-1559 minimumBaseFee mismatch!");
  }

  // Gas config confirmation
  if (bridgeData.data.gasConfig.gasPerCall !== BigInt(gasConfig.gasPerCall)) {
    throw new Error("Gas config gasPerCall mismatch!");
  }
  if (
    bridgeData.data.gasConfig.gasCostScaler !== BigInt(gasConfig.gasCostScaler)
  ) {
    throw new Error("Gas config gasCostScaler mismatch!");
  }
  if (
    bridgeData.data.gasConfig.gasCostScalerDp !==
    BigInt(gasConfig.gasCostScalerDp)
  ) {
    throw new Error("Gas config gasCostScalerDp mismatch!");
  }
  if (bridgeData.data.gasConfig.gasFeeReceiver !== gasConfig.gasFeeReceiver) {
    throw new Error("Gas config gasFeeReceiver mismatch!");
  }

  // Protocol config confirmation
  if (
    bridgeData.data.protocolConfig.blockIntervalRequirement !==
    BigInt(protocolConfig.blockIntervalRequirement)
  ) {
    throw new Error("Protocol config blockIntervalRequirement mismatch!");
  }

  // Buffer config confirmation
  if (
    bridgeData.data.bufferConfig.maxCallBufferSize !==
    BigInt(bufferConfig.maxCallBufferSize)
  ) {
    throw new Error("Buffer config maxCallBufferSize mismatch!");
  }

  // Base Oracle config confirmation
  if (
    bridgeData.data.baseOracleConfig.threshold !== baseOracleConfig.threshold
  ) {
    throw new Error("Base oracle config threshold mismatch!");
  }
  if (
    bridgeData.data.baseOracleConfig.signerCount !==
    baseOracleConfig.signerCount
  ) {
    throw new Error("Base oracle config signerCount mismatch!");
  }
  if (
    bridgeData.data.baseOracleConfig.signers.length !==
    baseOracleConfig.signers.length
  ) {
    throw new Error("Base oracle config signer array length mismatch!");
  }
  for (let i = 0; i < baseOracleConfig.signers.length; i++) {
    const onchain = bridgeData.data.baseOracleConfig.signers[i];
    const expected = baseOracleConfig.signers[i];
    if (onchain === undefined || expected === undefined) {
      throw new Error(`Base oracle config signer missing! Index: ${i}`);
    }
    if (toHex(new Uint8Array(onchain)) !== toHex(expected as Uint8Array)) {
      throw new Error(`Base oracle config signer mismatch! Index: ${i}`);
    }
  }

  // Partner oracle config confirmation
  if (
    bridgeData.data.partnerOracleConfig.requiredThreshold !==
    partnerOracleConfig.requiredThreshold
  ) {
    throw new Error("Partner oracle config threshold mismatch!");
  }
}

main().catch((e) => {
  console.error("❌ Initialization failed:", e);
  process.exit(1);
});
